_posts/2021-06-22-type-phantoms-and-safe-apis.md

```diff
@@ -0,0 +1,712 @@
+    }
+  }
+```
+
+This can be expanded to check error types and produce different status codes based on that. An error
+object library like [boom](https://hapi.dev/module/boom/) could be useful as well.
+
+An interesting alternative is to use more functional style, and a `Result` / `Either` type.
+Languages like [F#](https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/results),
+[Haskell](https://hackage.haskell.org/package/base/docs/Data-Either.html)
+and [Rust](https://doc.rust-lang.org/std/result/enum.Result.html) use them extensively.
+
+Result types
+are [extremely useful in TypeScript](https://imhoff.blog/posts/using-results-in-typescript)
+as well, and a bare-bones implementation is easy:
+
+```typescript
+export type Result<Success, Failure> =
+  { ok: true; value: Success } | { ok: false; error: Failure }
+```
+
+One way to use them on the server-side would be to expand the `Handler` type:
+
+```typescript
+// lib/Server.ts
+
+export type Handler<EP extends Endpoint> =
+  (params: Params<EP>) => Promise<Result<EP['result'], string>>
+
+const createHandler = <EP extends Endpoint>(handler: Handler<EP>): RouteHandler<EP> =>
+  async (req, res) => {
+    const result = await handler({ path: req.params, body: req.body })
+
+    if (result.ok) {
+      res.send(result.value)
+    } else {
+      res.status(500).send(result.error)
+    }
+  }
+```
+
+Of course, the backend code could throw exceptions regardless, so still catching those as well would
+be useful.
+
+Result types are even more useful in client code, where it's easy to forget to handle an error case.
+If the API client always returns a `Result` value, the error handling code needs to be made
+explicit, and there's no way to miss an exception. Using a `Result` type on the client-side works
+similarly to the server code:
+
+```typescript
+// lib/Client.ts
+
+export type Handler<EP extends Endpoint> =
+  (params: EP['path'], body: EP['body']) => Promise<Result<EP['result'], string>>
+
+/* ... */
+  function createHandler<EP extends T[keyof T]>({ method, pattern }: Route<EP>) {
+    const handler: Handler<EP> = (params, data) => {
+      const url = getUrl(pattern, params)
+
+      return client.request<EP['result']>({ data, method, url })
+        .then(r => ({ ok: true, value: r.data }))
+        .catch(e => ({ ok: false, value: e?.response ?? 'Request failed' }))
+    }
+
+    return handler
+  }
+
+/* ... */
+```
+
+### Validating requests
+
+While strict types can guarantee that the request content is correct to a high degree, there's still
+need for validation. Perhaps some requests don't use the utilities, and malicious requests are
+always possible.
+
+We can use the existing type definitions to make sure that all requests pass through validation.
+Assuming we already have some validation functions, we can combine those with the API model. The
+utility types guarantee that there must be a correctly typed validation function for every endpoint:
```

Comment:
You already mentioned io-ts decoding earlier, but perhaps a line or two about [parsing vs validating](https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate/)? I confess that I haven't properly compared the approaches myself yet — but from what I've seen, libraries like [zod](https://github.com/colinhacks/zod) make parsing seem an effective option. (It's in [use through Blitz, for example](https://blitzjs.com/docs/resolver-server-utilities#resolver-zod).)

Side note: looks like the author of zod has been [pondering the topic of this post too](https://trpc.io/) 😄

Reply:
☝🏼 After sleeping and re-reading this... This is probably orthogonal and wouldn't contribute value to the post. My point was to try and bring up the difference between building the typed things from the input by the means of some kind of a structure or schema, rather than validating the input with a type guard.

Reply:
Hmm... perhaps the topic is indeed not for this post.

Not sure I can leave mentioning tRPC out now that I've learned of it though. :thinking:

Reply:
Added a note about tRPC.